provider/google.go

```diff
@@ -220,16 +228,29 @@ func (p *GoogleProvider) DeleteRecords(endpoints []*endpoint.Endpoint) error {
 func (p *GoogleProvider) ApplyChanges(changes *plan.Changes) error {
 	change := &dns.Change{}
 
-	change.Additions = append(change.Additions, newRecords(changes.Create)...)
+	change.Additions = append(change.Additions, p.newRecords(changes.Create)...)
 
-	change.Additions = append(change.Additions, newRecords(changes.UpdateNew)...)
-	change.Deletions = append(change.Deletions, newRecords(changes.UpdateOld)...)
+	change.Additions = append(change.Additions, p.newRecords(changes.UpdateNew)...)
+	change.Deletions = append(change.Deletions, p.newRecords(changes.UpdateOld)...)
 
-	change.Deletions = append(change.Deletions, newRecords(changes.Delete)...)
+	change.Deletions = append(change.Deletions, p.newRecords(changes.Delete)...)
 
 	return p.submitChange(change)
 }
 
+// newRecords returns a collection of RecordSets based on the given endpoints and domainFilter.
+func (p *GoogleProvider) newRecords(endpoints []*endpoint.Endpoint) []*dns.ResourceRecordSet {
+	records := []*dns.ResourceRecordSet{}
+
+	for _, endpoint := range endpoints {
+		if p.domainFilter.Match(endpoint.DNSName) {
```

Comment:
why this is needed ? this is to simply convert the records to Cloud DNS format. The actual matching against the most suitable zone will occur in submitChange method, making filtering here unnecessary

Reply:
> I figured letting the Provider filter the Records as early in the code as possible would increase the safety of the implementation and prevent these kinds of logical errors in the future. 

Incorrect logging was there because domain filter was added after the most basic implementation (without any filtering). With this PR logging in submitChanges should be fixed. I am generally not a big fan of having two different methods sharing responsibility/contributing to the same goal, but in this case it seems fine. We should have tests to catch logical errors (of course not logging), but not add additional filtering on higher levels only because can forget them somewhere down the call stack. 

> If this filter isn't added at a top level in ApplyChanges, it would probably be duplicated in the two for-loops in seperateChange @ 313 and 321.

Yes that's the way it is currently working in Google and other providers. If we filter records out before passing to `submitChanges` method, they would be going through filtering again (indirectly though) in the separateChange(). 

In general, this change looks okay to me, however it is better to change newRecords -> newFilteredRecords as u suggested

Reply:
interesting context 👍

I've renamed `newRecords` to `newFilteredRecords`.
I appreciate the careful review 🙂